substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JDKInitializationFeature.java

```diff
@@ -87,6 +96,8 @@ public void afterRegistration(AfterRegistrationAccess access) {
         rci.initializeAtBuildTime("java.awt.font.TextAttribute", "Required for sun.text.bidi.BidiBase.NumericShapings");
         rci.initializeAtBuildTime("java.awt.font.NumericShaper", "Required for sun.text.bidi.BidiBase.NumericShapings");
 
+        rci.initializeAtRunTime("sun.awt.shell.Win32ShellFolder2$KnownLibraries", "Calls Win32ShellFolder2.loadKnownFolders()");
```

Comment:
Hmm, who initialized this class in the first place?

Reply:
I don't know who does that, but I know that class must not be initialized at build time. I didn't find these classes by debugging, but rather through reading of the OpenJDK code.

Reply:
@vjovanov Should I resolve this conversation? Or do you need further analysis?

Reply:
This is not our job: Only the classes that we initialize we should also make exceptions for. That is why we can't put this line until we know who initialized it.

Reply:
Does that means that a class initialized by an application should not be marked as "initializeAtRunTime"?

If so, then how is an application developer supposed to know that their application led to initialization of `sun.awt.shell.Win32ShellFolder2$KnownLibraries`, for instance?

And is the application developer supposed to dig into the OpenJDK sources to figure out that they need to initialize this class at runtime?

Just to clarify, my intention here is to mark classes that, if needed by the application, need to be initialized at runtime. I also want to notice that I'm not using `rerun...()` functions because I could not find where their semantics are defined.

Reply:
The policy is, if we initialize somethig at build-time we need to address all the dependencies. If someone else does, then it is their job. That is why I ask, who initialized this class.

Reply:
> The policy is, if we initialize somethig at build-time we need to address all the dependencies. If someone else does, then it is their job.

So if a developer builds a native image of their application, and somehow causes `sun.awt.shell.Win32ShellFolder2$KnownLibraries` to be initialized at build time (because that is **the default**), then the developer needs to figure this out on their own?! This is an unrealistic expectation, because a Java developer should **not** need to analyze the Java standard library sources or disassemble the generated native image in order to figure out this bug.

Standard library classes that should never be initialized at build time are **known** to the developers of the Java standard library, and they should be marked so. The policy you're mentioning does not make sense in the context of a Java developer generating a native image of their application.

> That is why I ask, who initialized this class.

Again, no one did, and no one should initialize these classes at build time, ever.

Reply:
Yes they should, build-time initialization is a complex and dangerous feature; it should be used with caution. If one uses it, they should take full responsibility for it and initialize everything properly. 

By analogy, hundreds of classes in the JDK should be marked as run-time initialized. It is a great burden to mark all those classes, especially when there are changes in the JDK  between versions.

Reply:
> If one uses it...

How can one disable it, so that runtime initialization is the default? Are there any flags to control this behavior?